File: python/mirage/commands/builtin/databricks_volume/mkdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.databricks_volume.glob import resolve_glob
from mirage.core.databricks_volume.mkdir import mkdir as mkdir_core
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("mkdir",
         resource="databricks_volume",
         spec=SPECS["mkdir"],
         write=True)
async def mkdir(
    accessor: DatabricksVolumeAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    p: bool = False,
    v: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        raise ValueError("mkdir: missing operand")
    paths = await resolve_glob(accessor, paths, index)
    created: dict[str, bytes] = {}
    lines: list[str] = []
    for path in paths:
        await mkdir_core(accessor, path, index, parents=p)
        created[path.strip_prefix] = b""
        if v:
            lines.append(f"mkdir: created directory '{path.original}'")
    output = ("\n".join(lines) + "\n").encode() if lines else None
    return output, IOResult(writes=created)

File: typescript/packages/core/src/commands/builtin/databricks_volume/mkdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DatabricksVolumeAccessor } from '../../../accessor/databricks_volume.ts'
import { resolveGlob } from '../../../core/databricks_volume/glob.ts'
import { mkdir as dbxMkdir } from '../../../core/databricks_volume/mkdir.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'

const ENC = new TextEncoder()

async function mkdirCommand(
  accessor: DatabricksVolumeAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('mkdir: missing operand\n') })]
  }
  const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
  const verbose = opts.flags.v === true
  const lines: string[] = []
  const writes: Record<string, Uint8Array> = {}
  for (const path of resolved) {
    await dbxMkdir(accessor, path)
    writes[path.original] = new Uint8Array()
    if (verbose) lines.push(`mkdir: created directory '${path.original}'`)
  }
  const output: ByteSource | null = lines.length > 0 ? ENC.encode(lines.join('\n') + '\n') : null
  return [output, new IOResult({ writes })]
}

export const DATABRICKS_VOLUME_MKDIR = command({
  name: 'mkdir',
  resource: ResourceName.DATABRICKS_VOLUME,
  spec: specOf('mkdir'),
  fn: mkdirCommand,
  write: true,
})
